pilot/pkg/serviceregistry/kube/controller/multicluster.go

```diff
@@ -51,107 +53,147 @@ type kubeController struct {
-	kubectl.AppendServiceHandler(func(svc *model.Service, ev model.Event) { m.updateHandler(svc) })
+	kubeRegistry.AppendServiceHandler(func(svc *model.Service, ev model.Event) { m.updateHandler(svc) })
+
+	// TODO move instance cache out of registries
+	if m.serviceEntryStore != nil && features.EnableServiceEntrySelectPods {
+		// Add an instance handler in the kubernetes registry to notify service entry store about pod events
+		kubeRegistry.AppendWorkloadHandler(m.serviceEntryStore.WorkloadInstanceHandler)
+	}
+
+	if localCluster {
+		// TODO implement deduping in aggregate registry to allow multiple k8s registries to handle WorkloadEntry
+		if m.serviceEntryStore != nil && features.EnableK8SServiceSelectWorkloadEntries {
+			// Add an instance handler in the service entry store to notify kubernetes about workload entry events
+			m.serviceEntryStore.AppendWorkloadHandler(kubeRegistry.WorkloadInstanceHandler)
+		}
+	}
+
+	// TODO only create namespace controller and cert patch for remote clusters (no way to tell currently)
+	if m.serviceController.Running() {
+		go kubeRegistry.Run(stopCh)
+	}
+	if m.fetchCaRoot() != nil && (features.ExternalIstioD || features.CentralIstioD || localCluster) {
+		// TODO remove initNamespaceController (and probably need leader election here? how will that work with multi-primary?)
+		log.Infof("joining leader-election for %s in %s", leaderelection.NamespaceController, options.SystemNamespace)
+		go leaderelection.
+			NewLeaderElection(options.SystemNamespace, m.serverID, leaderelection.NamespaceController, client.Kube()).
+			AddRunFunction(func(leaderStop <-chan struct{}) {
+				log.Infof("starting namespace controller for cluster %s", clusterID)
+				nc := NewNamespaceController(m.fetchCaRoot, client)
+				// Start informers again. This fixes the case where informers for namespace do not start,
+				// as we create them only after acquiring the leader lock
+				// Note: stop here should be the overall pilot stop, NOT the leader election stop. We are
+				// basically lazy loading the informer, if we stop it when we lose the lock we will never
+				// recreate it again.
+				client.RunAndWait(stopCh)
+				nc.Run(leaderStop)
+			}).Run(stopCh)
+	}
 
-	go kubectl.Run(stopCh)
+	// Patch cert if a webhook config name is provided.
+	// This requires RBAC permissions - a low-priv Istiod should not attempt to patch but rely on
+	// operator or CI/CD
 	webhookConfigName := strings.ReplaceAll(validationWebhookConfigNameTemplate, validationWebhookConfigNameTemplateVar, m.secretNamespace)
-	if m.fetchCaRoot != nil {
-		nc := NewNamespaceController(m.fetchCaRoot, clients)
-		go nc.Run(stopCh)
-		go webhooks.PatchCertLoop(features.InjectionWebhookConfigName.Get(), webhookName, m.caBundlePath, clients.Kube(), stopCh)
-		valicationWebhookController := webhooks.CreateValidationWebhookController(clients, webhookConfigName,
+	if features.InjectionWebhookConfigName.Get() != "" && m.caBundlePath != "" && !localCluster {
```

Comment:
@stevenctl @hzxuzhonghu Should n't this patching happen only for centralIstiod or externalIstiod model? With out that condition, multi cluster installation with control plane in primary cluster is trying to patch webhook in secondary cluster. This is not used to be like that earlier. Is it intentional?

Reply:
We're trying to make our remote clusters be "istiodless", our integration tests are setup to run in this way. I think we need more work to do that properly, like specifying that the cluster is a remote by some annotation in the remote secret.

Reply:
Ok. So for now, what do you suggest? We are trying to upgrade to the new version and it is trying to patch. Should we provide the necessary permissions and let it patch? Is there any work around?

Reply:
We can probably guard this with `features.ExternalIstioD || features.CentralIstioD` for now, AFAIK nobody is focused on istiod-less remotes for 1.9, and it will probably require one of those flags anyway.

Reply:
Also, are you trying to upgrade a multi-cluster setup?

Reply:
Yes